Change layout to reference the previous sibling (#50)
pavpanchekha/emberfox: book/layout.md

These layout objects will be the nodes of a *layout tree*. So we'll

> want layout objects to have children, a parent, and an HTML element

Reviewer: ```suggestion
want layout objects to have references to their children, a parent, and an HTML element
```

> that they correspond to. Each layout object will also have a layout to

Reviewer: ```suggestion
that they correspond to. Each layout object will also have a reference to
```

    self.flush()
```

> Besides the references to other layout object, each layout object also

Reviewer: ```suggestion
Besides the references to other related layout objects, each layout object also
```

def has_block_children(self):
    for child in self.node.children:
        if isinstance(child, TextNode):

>             if not child.text.isspace():

Reviewer: Why did these two lines get removed?

Author: Thanks to there no longer being whitespace-only TextNodes, after changes in Ch4.

``` {.python indent=4}
def layout(self):
    if self.has_block_children():
        previous = None
        for child in self.node.children:

>             if isinstance(child, TextNode): continue

Reviewer: Not skipping TextNode any more?

Author: There no won't be whitespace-only TextNodes, which is what this was skipping

greedy: paragraph and headings and other things take up all the
horizontal space they can. So they use the parent's width:

``` {.python}
self.w = self.parent.w
```

This also means that a block layout object's horizontal position is
the same as its parent's:

``` {.python}
self.x = self.parent.x
```

The vertical position of a block layout object depends on whether it
is its parent's first child or not. First children start at the top of
the parent, later children start where the previous layout object
ended:

``` {.python}
if self.previous:
    self.y = self.previous.y + self.previous.h
else:
    self.y = self.parent.y
```

After the width and position is computed, we can lay out all the
children of a block layout:

``` {.python indent=8}
for child in self.children:
    child.layout()
```

Note that the call to the child's `layout` method not only asks it to
compute its size, but also to its own children to the layout tree and
recursively call `layout` on them.

Finally, once all the children are laid out, a height of the layout
object can be computed:

``` {.python}
self.h = sum([child.h for child in self.children])
```

> Note how important the order is here. A block layout's width depends

Reviewer: ```suggestion
Note how important the order is here. A block layout object's width depends
```

on the parent's width; so by symmetry the width must be computed
before laying out the children. The position is the same: it has to be
computed before recursing so it's already available when the children
compute their own positions. But the height works the opposite way:
the height of a block layout depends on the height of its children, so
it must be computed after recursing, so that the heights of the
children are available.

This *dependency ordering* plays a really important role in
understanding how layout works, especially as layout gets more
complicated to support more features and faster speed. [Chapter
10](reflow.md) will explore this topic more.

That settles `BlockLayout`; let's now work on `InlineLayout`. Its `x`,
`w`, and `y` are set the same way as for a `BlockLayout`, but its
height is computed based on its *y*-cursor `cy` instead of the height
of its children.

``` {.python}
class InlineLayout:
    def layout(self):
        self.x = self.parent.x
        self.w = self.parent.w

        if self.previous:

>             self.y = self.previous.y + self.previous.h

Reviewer: Where did self.y get set in the old code? Don't see any line deletions in your PR for that.

Author: In the previous version, `x` and `y` were set by one's parent, and since `InlineLayout` doesn't have any children (at this stage of the book) there isn't code for `y` to delete. Honestly that was super confusing.

Reviewer: Got it, makes sense. A nice improvement.

``` {.python}
class DocumentLayout:
    def layout(self):
        # ...

>         self.w = WIDTH - 2*HSTEP

Reviewer: Why the change here?

Author: This is unrelated to the others, actually, but without padding there seems to be some kind of border that Tk adds by default which obscures some letters, and students wondered if that was a bug in their code.

Author: The padding comes back next chapter anyway.

Reviewer: OK can you add a comment explaining what's going on?
